### DISCUSSION TITLE
Testing play framework controller witch uses custom action

### DISCUSSION THREAD

**Discussion by lovilovasz** [playframework/playframework#11456] [Q&A]
I want to unit-test a controller method and to do that I need to mock out every dependency.

This is my controller:

  ```
  class InvoiceController @Inject() (
        cc: ControllerComponents,
        authAction: AuthAction,
        dao: InvoicerDao
    ) extends AbstractController(cc) {
    
      def saveInvoice: Action[AnyContent] = authAction { implicit userRequest =>
        Ok("ok")
      }
    }
```

I made a custom action with ActionBuilder and it is injected in the constructor. I use this action to Authenticate users.
I want to mock this out completely.

My unit test looks like this:

  ```
  class InvoiceControllerSpec extends PlaySpec with GuiceOneAppPerSuite {
    
      val authMock: AuthAction = mock[AuthAction]
      val daoMock: InvoicerDao = mock[InvoicerDao]
    
      val underTest: InvoiceController = new InvoiceController(
        Helpers.stubControllerComponents(),
        authMock,
        daoMock
      )
    
      "InvoiceController" should {
        "should save invoice" in {
          val body =
            Source.fromResource("InvoiceControllerSaveInvoice.json").mkString.trim
    
          val request = FakeRequest(POST, "/saveInvoice")
            .withJsonBody(Json.toJson(body))
    
          val userResult = Future {
            Ok("Authorized.")
          }
          when(authMock.invokeBlock(any, any)).thenReturn(userResult)
    
          val result: Future[Result] = underTest.saveInvoice.apply(
            request
          )
    
          status(result) mustBe 200
        }
      }
    }
```

After running the test I get NullPointerException.

I'm using mockito and scala test plus.

What am I missing or doing wrong?

If you know a better way to test this, please help.

Thank you!

**Comment by jducoeur** [comment]
> What am I missing or doing wrong?

This isn't meant to be at all snarky, but "using mockito" is my gut response.  It's a very brittle tool to begin with, and has incredibly sharp edges when applied to Scala.  It's extremely weakly-typed and null-prone, so this sort of error happens all the time.

(When I started at a previous job, I lost literally *months* just fixing this sort of Mockito error -- it was pretty miserable.)

If you must do mocking, I recommend the somewhat-more-robust [ScalaMock](https://scalamock.org/) instead.

But in general, my rule of thumb is to work quite differently.  If you're interested in exploring a different way of operating, I have [a series of articles](https://jducoeur.medium.com/a-philosophy-of-testing-1-introduction-dc5b7b43a211) that outlines the testing approach that I've found works best for serious Scala services, based on 10 years of doing so.
